sljit_src/sljitNativeS390X.c

```diff
@@ -0,0 +1,2817 @@
+			FAIL_IF(push_inst(compiler, oilf(dst_r, src2w)));
+			break;
+		case SLJIT_XOR32:
+			FAIL_IF(push_inst(compiler, xilf(dst_r, src2w)));
+			break;
+		case SLJIT_AND32:
+			FAIL_IF(push_inst(compiler, nilf(dst_r, src2w)));
+			break;
+		default:
+			SLJIT_UNREACHABLE();
+		}
+	}
+	else if ((src2 & SLJIT_IMM) && have_op_3_imm(op, src2w)) {
+		abort(); /* TODO(mundaym): implement */
+	}
+	else if ((src2 & SLJIT_MEM) && (dst_r == src1_r)) {
+		/* most 32-bit instructions can only handle 12-bit immediate offsets */
+		int need_u12 = !have_ldisp() &&
+			(op & SLJIT_I32_OP) &&
+			(GET_OPCODE(op) != SLJIT_ADDC) &&
+			(GET_OPCODE(op) != SLJIT_SUBC);
+		struct addr mem;
+		if (need_u12)
+			FAIL_IF(make_addr_bx(compiler, &mem, src2, src2w, tmp1));
+		else
+			FAIL_IF(make_addr_bxy(compiler, &mem, src2, src2w, tmp1));
+
+		int can_u12 = is_u12(mem.offset) ? 1 : 0;
+		sljit_ins ins = 0;
+		switch (GET_OPCODE(op) | (op & SLJIT_I32_OP)) {
+		/* 64-bit ops */
+		#define EVAL(op) op(dst_r, mem.offset, mem.index, mem.base)
+		case SLJIT_ADD:
+			ins = signed_flags ? EVAL(ag) : EVAL(alg);
+			break;
+		case SLJIT_SUB:
+			ins = signed_flags ? EVAL(sg) : EVAL(slg);
+			break;
+		case SLJIT_ADDC:
+			ins = EVAL(alcg);
+			break;
+		case SLJIT_SUBC:
+			ins = EVAL(slbg);
+			break;
+		case SLJIT_MUL:
+			ins = EVAL(msg);
+			break;
+		case SLJIT_OR:
+			ins = EVAL(og);
+			break;
+		case SLJIT_XOR:
+			ins = EVAL(xg);
+			break;
+		case SLJIT_AND:
+			ins = EVAL(ng);
+			break;
+		/* 32-bit ops */
+		case SLJIT_ADD32:
+			ins = signed_flags ?
+				(can_u12 ? EVAL(a) : EVAL(ay)):
+				(can_u12 ? EVAL(al) : EVAL(aly));
+			break;
+		case SLJIT_SUB32:
+			ins = signed_flags ?
+				(can_u12 ? EVAL(s) : EVAL(sy)):
+				(can_u12 ? EVAL(sl) : EVAL(sly));
+			break;
+		case SLJIT_ADDC32:
+			ins = EVAL(alc);
+			break;
+		case SLJIT_SUBC32:
+			ins = EVAL(slb);
+			break;
+		case SLJIT_MUL32:
+			ins = can_u12 ? EVAL(ms) : EVAL(msy);
+			break;
+		case SLJIT_OR32:
+			ins = can_u12 ? EVAL(o) : EVAL(oy);
+			break;
+		case SLJIT_XOR32:
+			ins = can_u12 ? EVAL(x) : EVAL(xy);
+			break;
+		case SLJIT_AND32:
+			ins = can_u12 ? EVAL(n) : EVAL(ny);
+			break;
+		#undef EVAL
+		default:
+			SLJIT_UNREACHABLE();
+		}
+		FAIL_IF(push_inst(compiler, ins));
+	}
+	else {
+		sljit_gpr src2_r = FAST_IS_REG(src2) ? gpr(src2 & REG_MASK) : tmp1;
+		if (src2 & SLJIT_IMM) {
+			/* load src2 into register */
+			FAIL_IF(push_load_imm_inst(compiler, src2_r, src2w));
+		}
+		if (src2 & SLJIT_MEM) {
+			/* load src2 into register */
+			FAIL_IF(load_word(compiler, src2_r, src2, src2w, tmp1, op & SLJIT_I32_OP));
+		}
+		/* TODO(mundaym): distinct operand facility where needed */
+		if (src1_r != dst_r && src1_r != tmp0) {
+			FAIL_IF(push_inst(compiler, (op & SLJIT_I32_OP) ?
+				lr(tmp0, src1_r) : lgr(tmp0, src1_r)));
+			src1_r = tmp0;
+		}
+		sljit_ins ins = 0;
+		switch (GET_OPCODE(op) | (op & SLJIT_I32_OP)) {
+		/* 64-bit ops */
+		case SLJIT_ADD: ins = signed_flags ?
```

Comment:
Incorrect syntax. There are other similar cases below. Maybe a similar macro to WHEN could make these nice.

Reply:
not sure if nicer, so suggest different names or scope if not